glean_parser/templates/javascript_server.jinja2

```diff
@@ -0,0 +1,120 @@
+   *
+   * @param {string} applicationId - The application ID.
+   * @param {string} appDisplayVersion - The application display version.
+   * @param {string} channel - The channel.
+   */
+  constructor(applicationId, appDisplayVersion, channel, logger_options) {
+    this._applicationId = applicationId;
+    this._appDisplayVersion = appDisplayVersion;
+    this._channel = channel;
+
+    if (!_logger) {
+      // append '-glean' to `logger_options.app` to avoid collision with other loggers and double logging
+      logger_options.app = logger_options.app + '-glean';
+      // set the format to `heka` so messages are properly ingested and decoded
+      logger_options.fmt = 'heka';
+      _logger = mozlog(logger_options)();
+    }
+    this._log = _logger;
+  }
+  /**
+   * Record and submit a server event object.
+   * Event is logged using internal mozlog logger.
+   *
+   {% for metric_type, metrics in metrics_by_type.items() %}
+   {% for metric in metrics %}
+   * @param { {{-metric|js_metric_type-}} } {{ metric|metric_argument_name }} - {{ metric|metric_argument_description }}.
+   {% endfor %}
+   {% endfor %}
+   */
+  record({
+    {% for metric_type, metrics in metrics_by_type.items() %}
+    {% for metric in metrics %}
+    {{ metric|metric_argument_name }},
+    {% endfor %}
+    {% endfor %}
+  }) {
+    let timestamp = new Date().toISOString();
+    let eventPayload = {
+      metrics: {
+        {% for metric_type, metrics in metrics_by_type.items() %}
+        {{ metric_type }}: {
+          {% for metric in metrics %}
+          '{{ metric|metric_name }}': {{ metric|metric_argument_name }},
+          {% endfor %}
+        },
+        {% endfor %}
+      },
+      ping_info: {
+        seq: 0, // this is required, however doesn't seem to be useful in server context
+        start_time: timestamp,
+        end_time: timestamp,
+      },
+      // `Unknown` fields below are required in the Glean schema, however they are not useful in server context
+      client_info: {
+        telemetry_sdk_build: 'glean_parser v{{ parser_version }}',
+        first_run_date: 'Unknown',
+        os: 'Unknown',
+        os_version: 'Unknown',
+        architecture: 'Unknown',
+        app_build: 'Unknown',
+        app_display_version: this._appDisplayVersion,
+        app_channel: this._channel,
+      },
+    };
+    eventPayload = JSON.stringify(eventPayload);
+
+    // This is the message structure that Decoder expects: https://github.com/mozilla/gcp-ingestion/pull/2400
+    let ping = {
+      document_namespace: this._applicationId,
+      document_type: '{{ ping }}',
+      document_version: '1',
+      document_id: uuidv4(),
+      payload: eventPayload,
+    };
+
+    // this is similar to how FxA currently logs with mozlog: https://github.com/mozilla/fxa/blob/4c5c702a7fcbf6f8c6b1f175e9172cdd21471eac/packages/fxa-auth-server/lib/log.js#L289
+    this._log.info(GLEAN_EVENT_MOZLOG_TYPE, ping);
```

Comment:
How does this then get to our pipeline?

Reply:
Asking because the proposal discussed significantly easier generated code that doesn't strictly adhere to our json ping schema. there is was specified that some "GCP Cloud logging to bigquery" adapter would be used, meaning this never goes through the early parts of the pipeline.

Reply:
These logs will be delivered to PubSub via Cloud Logging Sink, processed by the modified Decoder ([PR opened](https://github.com/mozilla/gcp-ingestion/pull/2400)) which will deliver them to `structured-decoded` topic. Next steps are standard pipeline processing (BQ live tables -> stable tables).

Reply:
Thanks, that was the missing context.